Add prior to likelihood evaluator.

cmbiwer opened the pull request:
This PR adds the functionality to calculate a prior to the likelihood evaluators. If no prior is given during initialization, then its no different than what was already implemented.

This is just a rudimentary implementation so that the samplers (kombine or emcee) can work with the ```pycbc.inference``` modules.

The prior returns ```-numpy.inf``` if the parameters are invalid. If parameters are invalid it does not bother generating a waveform and just returns the value.

EDIT: The prior can be a callable class or a function.

cdcapano commented on pycbc/inference/prior.py:
```diff
@@ -0,0 +1,35 @@
+    """ Function that returns default value if user does not specify a prior.
+    """
+    return 0
+
```
I tested this and it works right now, but I don't think it'll be possible for the priors to just be functions. They priors will need to know the order and names of parameters when they are passed to them, like the waveform generator does. Better that they be classes, in which one of the class methods is .prior, which takes a list of parameters and returns a value. Could you define a BasePrior (or maybe call it ZeroPrior) that has a variable_args attribute, similar to BaseGenerator? Then you would zip these together with the input parameters, like the generate function. The base one could be the no prior version (so, always returns 0), with future priors inheriting from this and replacing with their own prior functions.

cmbiwer replied:
Hey Collin, I believe what you are suggesting is already in this PR. I believe what you're talking about is the ```PriorEvaulator``` class.

cmbiwer replied:
Also I don't think we intend for priors to be functions as well. But these generator classes really just return 1 value. If we use ```__call__``` for classes to return the value instead of class functions named ```generate``` or ```prior```, its more general.

cdcapano replied:
Oh, I see. I loaded the page and tested before the commit that added this. I'll take a look.

cdcapano commented on pycbc/inference/likelihood.py:
```diff
@@ -191,10 +203,15 @@ def loglikelihood(self, params):
             parameter space.
         """
+        # get prior
+        prior = self.prior(params)
```
Call self._prior here rather than self.prior, it'll be slightly faster.

cmbiwer commented:
Changed to use ```self._prior```.

cdcapano commented on pycbc/inference/prior.py:
```diff
@@ -0,0 +1,93 @@
+
+    Parameters
+    ----------
+    variable_names : list
```
Call this variable_args to be consistent with generator.py.

cdcapano commented on pycbc/inference/prior.py:
```diff
@@ -0,0 +1,93 @@
+                 params_max=None):
+
+        # store the names of the variable params
+        self.variable_params = variable_params
```
In the doc string you called it variable_names; anyway, call it variable_args.

cdcapano commented on pycbc/inference/prior.py:
```diff
@@ -0,0 +1,93 @@
+        # acceptable values then return -inf
+        params = numpy.array(params)
+        result = sum((params < self.params_min) | (params > self.params_max))
+        if result:
```
Don't do a separate results. Just do ``if ((params < self.params_min) | (params > self.params_max)).all():``

cdcapano commented on pycbc/inference/prior.py:
```diff
@@ -0,0 +1,93 @@
+    def flat(self, param):
+        """ Simple flat prior for a single parameter.
+        """
+        return 0.0
```
Let's not make these class methods. The different types of distributions should just be stand-alone functions which the class will call. I'll add more details on the PR comments.

cmbiwer commented:
Hey Collin, I've renamed the var to ```variable_args```, changed the if statement for checking the min and max values, and moved ```flat_prior``` out of the class.

cdcapano commented:
Hi Chris, I think the PriorEvaluator should be agnostic to the types of distributions used. You've done that somewhat, but, for instance, you may have priors that have no bounds. I checked out your branch and made some changes to make it more agnostic, and (hopefully) to make it easier to add new distributions in the future. See [https://github.com/cmbiwer/pycbc/pull/4](this pull request). If it looks good to use, please merge into your branch, then I'll merge this onto master.

cmbiwer commented:
Hey Collin, I've merged it but let me fix the conflict in ```pycbc/inference/__init__.py``` first.

cmbiwer commented:
Hey Collin, I've fixed the merge conflict caused by #820. If you're fine with this, then please merge.

cdcapano commented:
Ok, looks good... I'll merge as soon as the Travis test is done.
